### PR TITLE
Ngrx

### PR DESCRIPTION
**Pull request by joredudiaz**
## Summary

- Implement state store
- Add book to shopping cart
- Remove book from shopping cart

## Jira Story

https://trello.com/c/29IQZlgL

## Screenshot

state

### CODE REVIEW

File: wolox-books/src/app/components/modal/modal.component.scss [hunk 1]
```diff
@@ -0,0 +1,29 @@
+@import '../../../colors.scss';
+
+.modal {
+  background-color: rgba(0,0,0,0.5);
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.scss]: use color rabiables

File: wolox-books/src/app/components/modal/modal.component.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<div class="modal" [class.hidde]="!open">
+  <div class="modal-container">
+    <button type="button" (click)="close()"><img src="../../../assets/close.png" alt="close" class="close"></button>
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.html]: close image tag

File: wolox-books/src/app/components/modal/modal.component.html [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+    <ng-content></ng-content>
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.html:4]: empty ng-content

File: wolox-books/src/app/components/modal/modal.component.scss [hunk 4]
```diff
@@ -0,0 +1,29 @@
+@import '../../../colors.scss';
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.scss]: remove file extension

File: wolox-books/src/app/components/modal/modal.component.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  transition: all .5s;
+  width: 100vw;
+  z-index: 1;
+}
+
+.modal-container {
+  background-color: $white;
+  border-radius: 15px;
+  height: 600px;
+  margin: 150px auto;
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.scss]: align-self center

File: wolox-books/src/app/components/modal/modal.component.scss [hunk 6]
```diff
@@ -0,0 +1,29 @@
+@import '../../../colors.scss';
+
+.modal {
+  background-color: rgba(0,0,0,0.5);
+  height: 100vh;
+  left: 0;
+  position: fixed;
+  top: 0;
+  transition: all .5s;
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.scss]: 0.5

File: wolox-books/src/app/components/modal/modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { Component, Input, Output, EventEmitter } from '@angular/core';
+
+@Component({
+  selector: 'modal',
+  templateUrl: './modal.component.html',
+  styleUrls: ['./modal.component.scss']
+})
+export class ModalComponent {
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.ts]: remove extra line

File: wolox-books/src/app/components/modal/modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  @Input('open')
+  open: boolean
+
+  @Output() openChange = new EventEmitter();
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.ts]:
```suggestion
  @Input('open');
  @Output() openChange = new EventEmitter();
    open: boolean

```

File: wolox-books/src/app/components/modal/modal.component.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  constructor() {
+    this.open = false;
+  }
+
+  close() {
+    this.openChange.emit(false);
+    this.open = false;
+  }
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/modal/modal.component.ts]: remove line

File: wolox-books/src/app/components/navbar/navbar.component.html [hunk 10]
```diff
@@ -1,4 +1,13 @@
 <div class="nav">
   <img src="../../assets/logo-wolox.png" alt="wolox-logo" class="logo">
-  <button class="nav-title" (click)="closeSession()">Logout</button>
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/navbar/navbar.component.html]: remove line

File: wolox-books/src/app/components/navbar/navbar.component.html [hunk 11]
```diff
@@ -1,4 +1,13 @@
...
+  <div class="left-container">
+    <div class="shopping-cart-container">
+      <button type="button" (click)="openModal()">
+        <img src="../../assets/shopping-cart.png" alt="shopping-cart" class="shopping-cart">
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/navbar/navbar.component.html]: close img tag

File: wolox-books/src/app/components/navbar/navbar.component.scss [hunk 12]
```diff
@@ -23,3 +22,30 @@
...
   line-height: 21px;
 }
+
+.left-container {
+  display: flex;
+}
+
+.shopping-cart {
+  height: 24px;
+  width: 22.31px;
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/navbar/navbar.component.scss]: remove dcimals

File: wolox-books/src/app/components/navbar/navbar.component.scss [hunk 13]
```diff
@@ -23,3 +22,30 @@
...
+
+.sup-quantity {
+  background-color: $cerulean;
+  border-radius: 50%;
+  bottom: 15px;
+  color: $white;
+  font-size: 10px;
+  height: 15px;
+  left: 10px;
+  padding: 2px 5px 2px 5px;
```
Comment (FrankIglesias) [inline wolox-books/src/app/components/navbar/navbar.component.scss]:
```suggestion
  padding: 2px 5px;
```

File: wolox-books/src/app/screens/book/book.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export interface Book {
+  id: number;
+  author: string;
+  title: string;
+  image: object;
+  publisher: string;
+  year: string;
+  genre: string;
+}
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/book/book.ts]: missing eol here

File: wolox-books/src/app/screens/book/components/list/list.component.html [hunk 15]
```diff
@@ -1,11 +1,14 @@
...
 </div>
-<div *ngFor="let book of books" class="book-container">
+
+<div *ngFor="let book of books | async" class="book-container">
   <div class="book-card" *ngFor="let number of [0,1,2,3,4,5]">
     <img src="../../../../../assets/book-cover.png" alt="" class="book-cover">
     <p class="book-title">{{book.title}}</p>
     <p class="book-author">{{book.author}}</p>
-    <button type="button" class="add"><img src="../../../../../assets/add.png" alt=""></button>
+    <button type="button" class="add"><img src="../../../../../assets/add.png" alt="" (click)="addShoppingCart(book)"></button>
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/book/components/list/list.component.html]: set the on click function to the button instead

File: wolox-books/src/app/screens/book/components/list/list.component.html [hunk 16]
```diff
@@ -1,11 +1,14 @@
...
   </div>
 </div>
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/book/components/list/list.component.html]: remove line

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html [hunk 17]
```diff
@@ -0,0 +1,17 @@
+<modal [(open)]="open"> 
+  <h1 class="modal-title">Tu carrito de compras</h1>
+  <div *ngFor="let book of books" class="modal-container">
+    <p class="modal-text">{{book.title}} - {{book.author}}</p>
+    <button type="button" (click)="deleteFromShoppingCart(book.id)">
+      <img src="../../../assets/delete.png" alt="delete" class="modal-delete">
+    </button>
+  </div>
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.html]: remove line

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+  <div *ngIf="books.length == 0" class="no-items">
+    <img src="../../../../../assets/no-items.png" alt="no-items" class="empty-shopping-cart">
+    <div class="empty-text">
+      <h1 class="empty-title">EL CARRITO EST&Aacute; VAC&Iacute;O</h1>
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.html]: This is not an h1

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.scss [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+  font-size: 22px;
+  height: 22px;
+  line-height: 26px;
+  margin-bottom: 30px;
+  width: 400px;
+}
+
+.modal-delete {
+  height: 24px;
+  width: 19.49px;
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.scss]: remove decimals

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.scss [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+  font-size: 32px;
+  font-weight: 700;
+  line-height: 38px;
+}
+
+.empty-description {
+  font-size: 24px;
+  line-height: 28px;
+  margin-top: 24px;
+}
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.scss]: add EOL

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+@Component({
+  selector: 'shopping-cart-list',
+  templateUrl: './list.component.html',
+  styleUrls: ['./list.component.scss']
+})
+export class ListComponent {
+
+  books: Book[];
+
+  _open: boolean;
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:16]: why this?

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+@Component({
+  selector: 'shopping-cart-list',
+  templateUrl: './list.component.html',
+  styleUrls: ['./list.component.scss']
+})
+export class ListComponent {
+
+  books: Book[];
+
```
Comment (FrankIglesias) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:15]: remove line

File: wolox-books/src/app/store/shopping-cart.reducer.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+const initialState = {
+  books: [],
+  modal: false
+}
+
+export function shoppingCartReducer(state = initialState, action: ShoppingCartActions.Actions ) {
+  switch (action.type) {
+    case ShoppingCartActions.REMOVE_SHOPPING_CART:
+      return Object.assign({}, state, { books: state.books.filter(element => element.id != action.payload)});
```
Comment (FrankIglesias) [inline wolox-books/src/app/store/shopping-cart.reducer.ts]: use destructuring

File: wolox-books/src/app/store/shopping-cart.reducer.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+    case ShoppingCartActions.ADD_SHOPPING_CART:
+      return Object.assign({}, state, { books: [...state.books, action.payload]});
```
Comment (FrankIglesias) [inline wolox-books/src/app/store/shopping-cart.reducer.ts]:
```suggestion
      return {...state, books: [...state.books, action.payload] }
```

**Review by FrankIglesias — CHANGES_REQUESTED** [review]

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html [hunk 25]
```diff
@@ -0,0 +1,17 @@
+<modal [(open)]="open"> 
+  <h1 class="modal-title">Tu carrito de compras</h1>
+  <div *ngFor="let book of books" class="modal-container">
+    <p class="modal-text">{{book.title}} - {{book.author}}</p>
+    <button type="button" (click)="deleteFromShoppingCart(book.id)">
+      <img src="../../../assets/delete.png" alt="delete" class="modal-delete">
+    </button>
+  </div>
+
+  <div *ngIf="books.length == 0" class="no-items">
```
Comment (betoobernal) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.html]: use strict equal or `!books.length`

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html [hunk 26]
```diff
@@ -0,0 +1,17 @@
+<modal [(open)]="open">
```
Comment (betoobernal) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.html:1]: NICE! 👍

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html [hunk 27]
```diff
@@ -0,0 +1,17 @@
+<modal [(open)]="open"> 
+  <h1 class="modal-title">Tu carrito de compras</h1>
+  <div *ngFor="let book of books" class="modal-container">
+    <p class="modal-text">{{book.title}} - {{book.author}}</p>
+    <button type="button" (click)="deleteFromShoppingCart(book.id)">
+      <img src="../../../assets/delete.png" alt="delete" class="modal-delete">
```
Comment (betoobernal) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.html]: close tag

File: wolox-books/src/app/screens/book/components/list/list.component.html (hunk 15 above)
Comment (betoobernal) [inline wolox-books/src/app/screens/book/components/list/list.component.html]: close img tag

File: wolox-books/src/app/components/navbar/navbar.component.ts [hunk 28]
```diff
@@ -9,10 +14,20 @@ import { Router } from '@angular/router';
 })
 export class NavbarComponent {
 
-  constructor(private sessionService: SessionService, private router: Router) { }
+  quantity: number;
+
+  constructor(private sessionService: SessionService, private router: Router, private store: Store<AppState>) {
+    this.store.select("shoppingCart").subscribe(element => {
+      this.quantity = element['books'].length || 0;
+    });
```
Comment (betoobernal) [inline wolox-books/src/app/components/navbar/navbar.component.ts]: Si crees que la propiedad `books` existe y siempre tiene valor, entonces es suficiente haciendo `element['books'].length`, pero esto no valida que la propiedad exista o que tenga valor (si **no existe valor** tu code actual arrojará un *TypeError*); si piensas que el segundo escenario puede ocurrir, entonces debes pensar en algo como esto:

```suggestion
    this.store.select("shoppingCart").subscribe(({ books }) => {
      this.quantity = books && books.length ? books.length : 0;
    });
```

File: wolox-books/src/app/components/modal/modal.component.html (hunk 3 above)
Comment (joredudiaz) [reply wolox-books/src/app/components/modal/modal.component.html:4]: (?)

File: wolox-books/src/app/components/navbar/navbar.component.html (hunk 11 above)
Comment (joredudiaz) [reply wolox-books/src/app/components/navbar/navbar.component.html]: ERROR in Errors parsing template: Void elements do not have end tags "img" ("tton type="button" (click)="close()">close[ERROR ->]</img></button>
    <ng-content></ng-content>
  </div>
"): /Users/jorgediaz/Documents/projects/trainings/jd-angular/wolox-books/src/app/components/modal/modal.component.html@2:107

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.html (hunk 27 above)
Comment (joredudiaz) [reply wolox-books/src/app/screens/shopping-cart/components/list/list.component.html]: ERROR in Errors parsing template: Void elements do not have end tags "img" ("tems">
    no-items[ERROR ->]</img>
    <div class="empty-text">
      <p class="empty-title">EL CARRITO EST&Aacute; VAC&Iacute;O<"): /Users/jorgediaz/Documents/projects/trainings/jd-angular/wolox-books/src/app/screens/shopping-cart/components/list/list.component.html@9:93

File: wolox-books/src/app/screens/book/components/list/list.component.html (hunk 15 above)
Comment (joredudiaz) [reply wolox-books/src/app/screens/book/components/list/list.component.html]: ERROR in Errors parsing template: Void elements do not have end tags "img" ("tems">
    no-items[ERROR ->]</img>
    <div class="empty-text">
      <p class="empty-title">EL CARRITO EST&Aacute; VAC&Iacute;O<"): /Users/jorgediaz/Documents/projects/trainings/jd-angular/wolox-books/src/app/screens/shopping-cart/components/list/list.component.html@9:93

File: wolox-books/src/app/components/modal/modal.component.ts (hunk 8 above)
Comment (joredudiaz) [reply wolox-books/src/app/components/modal/modal.component.ts]: Why do you go down the `open` variable?

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts (hunk 21 above)
Comment (joredudiaz) [reply wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:16]: acts like a two way binding state, `_open` sets the initial state on constructor that the get won't be pass if I don't get the value from the observable that must be more complex the methods.

File: wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+  styleUrls: ['./list.component.scss']
+})
+export class ListComponent {
+  books: Book[];
+
+  _open: boolean;
+
+  constructor(private store: Store<AppState>) {
+    this.store.select('shoppingCart').subscribe(element => {
+      this._open = element['modal'];
```
Comment (betoobernal) [inline wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:20]: is necessary `_open`? a glance in the template tells me that `this.open` should be the same. If you want to show me the converse, cool.

Comment (joredudiaz) [reply wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:20]: Es necesario por que `_open` guarda el estado que se devuelve en `get open`, lo hago asi por que aunque este utilizando el estado de ngrx, para devolver el estado necesito un boolean no un Observable que es lo que se retornaria de llamar el store. https://github.com/wolox-training/jd-angular/pull/19#discussion_r441200187

Comment (betoobernal) [reply wolox-books/src/app/screens/shopping-cart/components/list/list.component.ts:20]: Ohh, okay. now i catch it

**Review by betoobernal — APPROVED** [review]

**Review by FrankIglesias — APPROVED** [review]

**Pull request merged**